File: backend/schemas/data_test.py
from __future__ import annotations

from pydantic import BaseModel


class DataTestRequest(BaseModel):
    symbols: list[str]


class DataTestField(BaseModel):
    key: str
    label: str


SOURCE_NAMES = ("interactive_service", "yahoo_finance", "finnhub", "fmp")

# {symbol: {field_key: {source_name: value | None}}}
DataTestValues = dict[str, dict[str, dict[str, float | str | None]]]


class DataTestResult(BaseModel):
    fields: list[DataTestField]
    symbols: list[str]
    values: DataTestValues
    source_errors: dict[str, str | None]

File: frontend/src/types/dataTest.ts
export interface DataTestField {
  key: string
  label: string
}

export type DataTestSourceName = 'interactive_service' | 'yahoo_finance' | 'finnhub' | 'fmp'

export type DataTestCellValues = Partial<Record<DataTestSourceName, number | string | null>>

export interface DataTestResult {
  fields: DataTestField[]
  symbols: string[]
  values: Record<string, Record<string, DataTestCellValues>>
  source_errors: Partial<Record<DataTestSourceName, string | null>>
}
